utils/man/energymon-file-provider.1
.TH "ENERGYMON\-FILE\-PROVIDER" "1" "2018-11-04" "energymon" "EnergyMon Utilities"
.SH "NAME"
.LP
energymon\-file\-provider \- write energy readings to a file at regular
intervals
.SH "SYNPOSIS"
.LP
\fBenergymon\-file\-provider\fP
[\fIOPTION\fP]... [\fIFILE\fP]
.SH "DESCRIPTION"
.LP
Writes energy readings to a file at regular intervals.
Uses the default EnergyMon implementation.
.LP
If no additional options are specified, energy readings are written to standard
output at the implementation's minimum update interval, and the program loops
until interrupted with \fBCTRL\-C\fP.
Note that using the default interval is not always desirable, as it may be too
fast and cause unnecessary overhead.
The data also tends to be quite noisy at the minimum update interval.
.LP
If \fIFILE\fP is specified, the file is overwritten with each write, so there
is only ever a single value.
To instead write a new line for each sample, specify \fI\-n/\-\-no\-rewind\fP.
This option is implied when using standard output by default.
.SH "OPTIONS"
.LP
.TP
\fB\-h\fP, \fB\-\-help\fP
Prints the help screen.
.TP
\fB\-c\fP, \fB\-\-count\fP=\fIN\fP
Stop after \fIN\fP reads.
.TP
\fB\-i\fP, \fB\-\-interval=\fP\fIUS\fP
The update interval in microseconds.
Must be greater than 0.
.TP
\fB\-n\fP, \fB\-\-no\-rewind\fP
Write each reading on a new line.
.SH "EXAMPLES"
.TP
\fBenergymon\-file\-provider\fP
Print energy values to standard output at the implementation's update interval.
.TP
\fBenergymon\-file\-provider energy.log\fP
Write energy readings to \fBenergy.log\fP at the implementation's update
interval, overwriting prior energy readings.
.TP
\fBenergymon\-file\-provider \-i 5000000 \-n energy.log\fP
Write energy readings to \fBenergy.log\fP every 5 seconds, with a new line for
each energy reading.
.TP
\fBenergymon\-file\-provider \-c 10 \-i 1000000\fP
Print energy values to standard output every second, stopping after 10
readings.
.SH "BUGS"
.LP
Report bugs upstream at <https://github.com/energymon/energymon>
.SH "SEE ALSO"
.BR energymon\-power\-poller (1)

RELEASES.md
# Release Notes

## [Unreleased]
### Changed
 * More pedantic formatting in man pages

## [v0.2.4] - 2018-06-01
### Changed
 * WattsUp implementations:
   * Wait up to 5 seconds for a good data packet (or fail) before completing initialization
   * Purge data buffers during initialization
   * Poll device at shorter intervals to increase energy value accuracy
   * Better handling of incomplete data packets
 * osp-polling: Allow polling thread to be interrupted during sleep when finish is requested
 * energymon-power-poller: Account for true elapsed time to produce more accurate power values

### Fixed
 * wattsup-libftdi, wattsup-libusb: Fix empty reads by always setting baud rate and serial line properties
 * wattsup-dev: Use pselect to avoid timer expired errors
 * odroid, odroid-ioctl, osp-polling, zcu102: Minor correction to polling delay interval
 * msr: Possible (but unlikely) precision computation bug
 * osp, osp-polling: Fix status check during init that didn't always work


## [v0.2.3] - 2017-11-19
### Added
 * Implementation for Xilinx ZCU102 (Jason Miller)

### Fixed
 * Fixed [#40]: Time conversion bug in common function 'energymon_gettime_elapsed_us' (Jason Miller)


## [v0.2.2] - 2017-10-26
### Added
 * New capabilities to some utility applications:
   * energymon-cmd-profile: Accept commands without quotations; add help option
   * energymon-file-provider: Add help, count, interval, and no-rewind options
   * energymon-idle-power: Add help option
 * Man pages for utility applications
 * VERSION and SOVERSION properties on shared object libraries
 * Multiarch support (use GNU standard installation directories)

### Changed
 * Increased minimum CMake version from 2.8 to 2.8.5 to support GNUInstallDirs
 * Refactor this RELEASES.md file

### Removed
 * Removed private symbol exports in shared object libraries (prevents symbol name clashes)

### Fixed
 * Various bug fixes to utility applications


## [v0.2.1] - 2017-10-01
### Added
 * Support for msr-safe kernel module in msr implementation
 * New cray-pm implementations to support capabilities of Cray XC30 and XC40 systems
 * New wattsup-libusb and wattsup-libftdi implementations
 * AppVeyor test configuration for Windows continuous integration

### Changed
 * Faster refresh in osp-polling, odroid and odroid-ioctl implementations
 * Better multi-platform support for some implementations, especially with portable time and sleep abstractions
 * Downgraded required CMake version from 2.8.9 to 2.8
 * Documented FSE 2016 and Technical Report publication information
 * Various enhancements, optimizations, and documentation improvements

### Deprecated
 * Deprecated custom DEFAULT CMake option in favor of ENERGYMON_BUILD_DEFAULT

### Removed
 * Removed Android CMake toolchain file: cmake-toolchain/android.toolchain.cmake

### Fixed
 * Fixed [#37]: In osp implementation, force overflow in ODROID Smart Power at 1000 Wh to avoid device counter saturation at 8192 Wh
 * Fixed [#27]: More modular build process
 * Numerous other bugs

## v0.2.0 - 2016-10-12
### Added
 * Initial public release

[Unreleased]: https://github.com/energymon/energymon/compare/v0.2.4...HEAD
[v0.2.4]: https://github.com/energymon/energymon/compare/v0.2.3...v0.2.4
[v0.2.3]: https://github.com/energymon/energymon/compare/v0.2.2...v0.2.3
[v0.2.2]: https://github.com/energymon/energymon/compare/v0.2.1...v0.2.2
[v0.2.1]: https://github.com/energymon/energymon/compare/v0.2.0...v0.2.1
[#40]: https://github.com/energymon/energymon/issues/40
[#37]: https://github.com/energymon/energymon/issues/37
[#27]: https://github.com/energymon/energymon/issues/27


utils/energymon-file-provider.c
/**
 * Polls an energymon at the provided interval and writes the output to the
 * file specified.
 *
 * @author Connor Imes
 * @date 2015-07-09
 */
#define _GNU_SOURCE
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <limits.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include "energymon-default.h"
#include "energymon-time-util.h"

static volatile int running = 1;
static int count = 0;
static const char* filename = NULL;
static uint64_t interval = 0;
static int is_rewind = 1;

static const char short_options[] = "+hc:i:n";
static const struct option long_options[] = {
  {"help",      no_argument,       NULL, 'h'},
  {"count",     required_argument, NULL, 'c'},
  {"interval",  required_argument, NULL, 'i'},
  {"no-rewind", no_argument,       NULL, 'n'},
  {0, 0, 0, 0}
};

static void print_usage(int exit_code) {
  fprintf(exit_code ? stderr : stdout,
          "Usage: energymon-file-provider [OPTION]... [FILE]\n"
          "Options:\n"
          "  -h, --help               Print this message and exit\n"
          "  -c, --count=N            Stop after N reads\n"
          "  -i, --interval=US        The update interval in microseconds\n"
          "  -n, --no-rewind          Write each reading on a new line\n");
  exit(exit_code);
}

static void parse_args(int argc, char** argv) {
  int c;
  while ((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
    switch (c) {
      case 'h':
        print_usage(0);
        break;
      case 'c':
        count = 1;
        running = strtoull(optarg, NULL, 0);
        break;
      case 'i':
        interval = strtoull(optarg, NULL, 0);
        if (interval == 0) {
          fprintf(stderr, "Interval must be > 0\n");
          print_usage(1);
        }
        break;
      case 'n':
        is_rewind = 0;
        break;
      case '?':
      default:
        print_usage(1);
        break;
    }
  }
  // get the file name (if set)
  if (optind < argc) {
    filename = argv[optind];
  }
}

static void shandle(int sig) {
  switch (sig) {
    case SIGTERM:
    case SIGINT:
#ifdef SIGQUIT
    case SIGQUIT:
#endif
#ifdef SIGKILL
    case SIGKILL:
#endif
#ifdef SIGHUP
    case SIGHUP:
#endif
      running = 0;
    default:
      break;
  }
}

int main(int argc, char** argv) {
  energymon em;
  uint64_t min_interval;
  uint64_t energy;
  FILE* fout = stdout;
  int ret = 0;

  // register the signal handler
  signal(SIGINT, shandle);

  parse_args(argc, argv);

  // initialize
  if (energymon_get_default(&em)) {
    perror("energymon_get_default");
    return 1;
  }
  if (em.finit(&em)) {
    perror("energymon:finit");
    return 1;
  }

  // get the update interval
  min_interval = em.finterval(&em);
  if (interval == 0) {
    interval = min_interval;
  } else if (interval < min_interval) {
    fprintf(stderr,
            "Requested interval is too short, minimum available: %"PRIu64"\n",
            min_interval);
    em.ffinish(&em);
    return 1;
  }

  // get the file (if set)
  if (filename != NULL) {
    fout = fopen(filename, "w");
    if (fout == NULL) {
      perror(filename);
      return 1;
    }
  }

  // update file at regular intervals
  while (running) {
    if (count) {
      running--;
    }
    // While the interface specifies that energy values are monotonically
    // increasing, in practice that may not always hold.
    // We need to guarantee that we don't leave trailing characters if later
    // readings are a smaller order of magnitude than earlier ones.
    // Reopening the file seems to be the best way to do this.
    if (is_rewind) {
      fout = freopen(NULL, "w", fout);
      if (fout == NULL) {
        perror("Reopening output file");
        ret = 1;
        break;
      }
    }
    errno = 0;
    energy = em.fread(&em);
    if (!energy && errno) {
      perror("energymon:fread");
      // continue anyway and hope the error is transitive
      // ret = 1;
      // break;
    }
    if (fprintf(fout, "%"PRIu64"\n", energy) < 0) {
      perror("Writing to output file");
      ret = 1;
      break;
    }
    fflush(fout);
    if (running) {
      energymon_sleep_us(interval, &running);
    }
  }

  // cleanup
  if (em.ffinish(&em)) {
    perror("energymon:ffinish");
  }
  if (fout != NULL && fclose(fout)) {
    perror("Closing output file");
  }

  return ret;
}
